pkg/kubefedctl/orphaning/orphaning.go

```diff
@@ -0,0 +1,133 @@
+	flags.StringVarP(&o.resourceNamespace, "namespace", "n", "", "If present, the namespace scope for this CLI request")
+	err := flags.MarkHidden("kubefed-namespace")
+	if err != nil {
+		return err
+	}
+	err = flags.MarkHidden("dry-run")
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// NewCmdOrphaning the head of orphaning-deletion sub commands
+func NewCmdOrphaning(cmdOut io.Writer, config util.FedConfig) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "orphaning-deletion",
+		Short: "Manage orphaning delete policy",
+		Long:  "Manage orphaning delete policy",
+		Run: func(cmd *cobra.Command, args []string) {
+			err := cmd.Help()
+			if err != nil {
+				klog.Fatalf("Error: %v", err)
+			}
+		},
+	}
+	cmd.AddCommand(newCmdEnableOrphaning(cmdOut, config))
+	cmd.AddCommand(newCmdDisableOrphaning(cmdOut, config))
+	cmd.AddCommand(newCmdStatusOrphaning(cmdOut, config))
+
+	return cmd
+}
+
+//  Complete ensures that options are valid and marshals them if necessary.
+func (o *orphanResource) Complete(args []string) error {
+	if len(args) == 0 {
+		return errors.New("resource type is required")
+	}
+
+	o.typeName = args[0]
+
+	if len(args) == 1 {
+		return errors.New("resource name is required")
+	}
+	o.resourceName = args[1]
+
+	if len(o.resourceNamespace) == 0 {
+		o.resourceNamespace = "default"
+	}
+
+	return nil
+}
+
+// Returns a Federated Resources Interface
+func (o *orphanResource) GetResourceClient(config util.FedConfig) (dynamic.ResourceInterface, error) {
+	hostConfig, err := config.HostConfig(o.HostClusterContext, o.Kubeconfig)
+	if err != nil {
+		return nil, errors.Wrapf(err, "Failed to get HostConfig for host cluster context '%s' and kubeconfig '%s'",
+			o.HostClusterContext, o.Kubeconfig)
+	}
+	// Lookup kubernetes API availability
+	apiResource, err := enable.LookupAPIResource(hostConfig, o.typeName, "")
+	if err != nil {
+		return nil, errors.Wrapf(err, "Failed to find targeted %s type", o.typeName)
+	}
+	klog.V(2).Infof("API Resource for %s.%s/%s found", apiResource.Name, apiResource.Group, apiResource.Version)
+	if !util.IsFederatedAPIResource(apiResource.Kind, apiResource.Group) {
+		return nil, errors.Errorf("Looks like the resource is not a Federated object, Kind='%v' Group='%v'",
+			apiResource.Kind, apiResource.Group)
+	}
+	targetClient, err := ctlutil.NewResourceClient(hostConfig, apiResource)
+
+	if err != nil {
+		return nil, errors.Wrapf(err, "Error creating client for %s", apiResource.Kind)
+	}
+
+	resourceClient := targetClient.Resources(o.resourceNamespace)
+	return resourceClient, nil
+}
+
+// Returns the Federated resource where the orphaning-deletion will be managed
+func (o *orphanResource) GetFederatedResource(resourceClient dynamic.ResourceInterface) (*unstructured.Unstructured, error) {
+	resource, err := resourceClient.Get(o.resourceName, metav1.GetOptions{})
+	if err != nil {
+		return nil, errors.Wrapf(err, "Error retrieving resource %s %q", o.typeName,
```

Comment:
Would it be possible to use the kind here rather than the type name the user passed in?

Reply:
Due to separation of `GetResourceClient` and `GetFederatedResource`, it will be complicated. In the `GetFederatedResource` I don't have a reference to `resourceClient` only to the `dynamic.ResourceInterface`.
In order to use Kind, I have to 

- or return `Kind` from  `GetResourceClient`  and provide it to `GetFederatedResource`
- or move `targetClient.Resources(o.resourceNamespace)` out of `GetResourceClient`.

Both options look problematic for me. What do you think ?

Reply:
I think the first option is a change worth making to ensure the UX consistency.

Reply:
I implemented the first option. 
For generalizing, `GetResourceClient` returns entire `*metav1.APIResource`, but it can be easily replace by `Kind` only.  
However, there is an additional option: the current error message looks like 
`Error: Error retrieving resource FederatedConfigMap "default/game-config1": federatedconfigmaps.types.kubefed.k8s.io "game-config1" not found`
 
Do we really have to print the resource `Kind` or its user provided type? Anyway the dynamic client prints (at list in the most common case) the resource name and group. Maybe we should just return the dynamic client error.
What do you think?